iree/samples/rt/vulkan/vk_graphics_integration.cc

```diff
@@ -70,15 +72,102 @@ static void check_vk_result(VkResult err) {
+                                 iree_required_extensions,
+                                 &iree_required_extensions_count);
+
+  // Ask IREE for its list of optional instance extensions.
+  iree_host_size_t iree_optional_extensions_count = 0;
+  iree_hal_vulkan_get_extensions(IREE_HAL_VULKAN_INSTANCE_OPTIONAL, options, 0,
+                                 NULL, &iree_optional_extensions_count);
+  const char** iree_optional_extensions =
+      new const char*[iree_optional_extensions_count];
+  iree_hal_vulkan_get_extensions(IREE_HAL_VULKAN_INSTANCE_OPTIONAL, options,
+                                 iree_optional_extensions_count,
+                                 iree_optional_extensions,
+                                 &iree_optional_extensions_count);
+
+  // Merge extensions lists, including optional and required for simplicity.
+  std::set<const char*> ext_set;
+  for (int i = 0; i < sdl_extensions_count; ++i) {
+    ext_set.insert(sdl_extensions[i]);
+  }
+  for (int i = 0; i < iree_required_extensions_count; ++i) {
+    ext_set.insert(iree_required_extensions[i]);
+  }
+  for (int i = 0; i < iree_optional_extensions_count; ++i) {
+    ext_set.insert(iree_optional_extensions[i]);
+  }
+  std::vector<const char*> extensions(ext_set.begin(), ext_set.end());
+  return extensions;
+}
+
+static std::vector<const char*> GetDeviceExtensions(
+    iree_hal_vulkan_extensibility_options_t options) {
+  // Ask IREE for its list of required device extensions.
+  iree_host_size_t iree_required_extensions_count = 0;
+  iree_hal_vulkan_get_extensions(IREE_HAL_VULKAN_DEVICE_REQUIRED, options, 0,
+                                 NULL, &iree_required_extensions_count);
+  const char** iree_required_extensions =
+      new const char*[iree_required_extensions_count];
+  iree_hal_vulkan_get_extensions(
+      IREE_HAL_VULKAN_DEVICE_REQUIRED, options, iree_required_extensions_count,
+      iree_required_extensions, &iree_required_extensions_count);
+
+  // Ask IREE for its list of optional device extensions.
+  iree_host_size_t iree_optional_extensions_count = 0;
+  iree_hal_vulkan_get_extensions(IREE_HAL_VULKAN_DEVICE_OPTIONAL, options, 0,
+                                 NULL, &iree_optional_extensions_count);
+  const char** iree_optional_extensions =
+      new const char*[iree_optional_extensions_count];
+  iree_hal_vulkan_get_extensions(
+      IREE_HAL_VULKAN_DEVICE_OPTIONAL, options, iree_optional_extensions_count,
+      iree_optional_extensions, &iree_optional_extensions_count);
+
+  // Merge extensions lists, including optional and required for simplicity.
+  std::set<const char*> ext_set;
+  ext_set.insert("VK_KHR_swapchain");
+  for (int i = 0; i < iree_required_extensions_count; ++i) {
+    ext_set.insert(iree_required_extensions[i]);
+  }
+  for (int i = 0; i < iree_optional_extensions_count; ++i) {
+    ext_set.insert(iree_optional_extensions[i]);
+  }
+  std::vector<const char*> extensions(ext_set.begin(), ext_set.end());
```

Comment:
std::vector<std::string> so that string lifetime is handled

Reply:
(may also be worth noting the lifetime of the extension strings on get_extensions; which is basically the lifetime of the driver)

Reply:
Hmm.. I'm not clear on the lifetime, and `ppEnabledExtensionNames` is still expecting char**

Reply:
lifetime is tied to the loader shared object or instance, depending on where they came from

add that note in the get_extensions and then I think this is fine